src/platform/avr/usb.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * SPDX-License-Identifier: MIT
+ * SPDX-FileCopyrightText: 2022 Rafael Silva <perigoso@riseup.net>
+ */
+
+#include <avr/interrupt.h>
+
+#include "protocol/reports.h"
+#include "usb.h"
+#include "util/hid_descriptors.h"
+#include "util/types.h"
+
+#include <USB.h>
+
+/** Buffer to hold the previously generated HID report, for comparison purposes inside the HID class driver. */
+static u8 openinput_hid_report_buff[sizeof(struct oi_report_t)];
+static u8 mouse_hid_report_buff[sizeof(struct mouse_report)];
+static u8 keyboard_hid_report_buff[sizeof(struct keyboard_report)];
+
+static u8 new_oi_report;
+static struct oi_report_t oi_report;
+static u8 oi_report_size;
+
+static u8 new_mouse_report;
+static struct mouse_report mouse_report;
+static u8 mouse_report_size;
+
+static u8 new_keyboard_report;
+static struct keyboard_report keyboard_report;
+static u8 keyboard_report_size;
+
+static struct protocol_config_t protocol_config;
+
+/** LUFA HID Class driver interface configuration and state information. This structure is
+ *  passed to all HID Class driver functions, so that multiple instances of the same class
+ *  within a device can be differentiated from one another.
+ */
+USB_ClassInfo_HID_Device_t openinput_hid_interface = {
```

Comment:
Is this aliased to `const`? I don't see changes to this memory so perhaps this can be on `.data` section or other read-only memory.

Reply:
This is just me playing with lufa and committing to save progress, this is will get a clean up

Reply:
The struct itself i don't think is written, so yes, although on avrs, consts are always fetched to ram anyway i believe

Reply:
> The struct itself i don't think is written, so yes, although on avrs, consts are always fetched to ram anyway i believe

Well, yeah, I'm not aware of those embedded cases. Godbolt https://godbolt.org/z/soo4eaaev tells me it goes to `.rodata` and the compiler can optimize it to immediate loads instead of an offset. This is not always the case although, as I believe the jump cost is a trade-off the compiler takes into consideration, depending on size of the data/non-natural alignment.